test/src/unit-regression2.cpp

```diff
@@ -835,6 +849,27 @@ TEST_CASE("regression tests 2")
 
         CHECK(j.dump() == "[1,4]");
     }
+
+    SECTION("issue #3215 - Exception thrown during initialization causes a memory leak")
+    {
+        std::string s;
+        try
+        {
+            // std::map<std::string, std::string> j // no leak
+            nlohmann::json j // leak
```

Comment:
This looks like the error will be during the construction of the `initializer_list` and not during the `json` constructor.   Does it reproduce with `nlohmann::json::initializer_list_t j`?

Reply:
I try: https://github.com/nlohmann/json/pull/3222/commits/66a2d18df6f70beabbd4924d12bf3d90c399635a

Reply:
Yes, the leak remains: https://github.com/nlohmann/json/runs/4694680374?check_suite_focus=true

Reply:
t looks to me like the compiler is constructing some of the initializer list data, then hitting the exception and not cleaning up what it already constructed.  I can't prove it at this point, but I don't see anything yet that says that the library is doing anything wrong.  I guess the next check would be `initializer_list<nlohmann::json>` to remove `json_ref` from the equation.

Reply:
I'll try. Would be strange if this was a compiler bug.

Reply:
Yes, strange, but it's a bit of an edge case.

Reply:
It fails for 

    std::initializer_list<nlohmann::json>

but passes for

    std::initializer_list<std::pair<std::string, std::string>>

Not sure yet what that means.  

https://godbolt.org/z/hef9Yf8P5

Reply:
Fails the same way for this too:

        nlohmann::json j{
            "smth", f_3215(true) // exception thrown here
        };